src/CacheClient.cpp

```diff
@@ -0,0 +1,125 @@
+#include "CacheClient.hpp"
+#include "Hash.hpp"
+#include "DagData.hpp"
+#include "RunAction.hpp"
+#include "NodeResultPrinting.hpp"
+#include "StatCache.hpp"
+#include "Profiler.hpp"
+#include "Exec.hpp"
+#include "BuildQueue.hpp"
+#include "MakeDirectories.hpp"
+
+const char* kENV_REAPI_CACHE_CLIENT = "REAPI_CACHE_CLIENT";
+const char* kENV_CACHE_SERVER_ADDRESS = "CACHE_SERVER_ADDRESS";
+const char* kENV_BEE_CACHE_BEHAVIOUR = "BEE_CACHE_BEHAVIOUR";
+
+static int SlowCallback(void *user_data, const char* label)
+{
+    SlowCallbackData *data = (SlowCallbackData *)user_data;
+    MutexLock(data->queue_lock);
+    char buffer[1000];
+    snprintf(buffer,sizeof(buffer),"%s %s", data->node_data->m_Annotation.Get(), label);
+    int sendNextCallbackIn = PrintNodeInProgress(data->node_data, data->time_of_start, data->build_queue, buffer);
+    MutexUnlock(data->queue_lock);
+    return sendNextCallbackIn;
+}
+
+static int SlowCallback_CacheRead(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheRead]");
+}
+
+static int SlowCallback_CacheWrite(void *user_data)
+{
+    return SlowCallback(user_data, "[CacheWrite]");
+}
+
+enum Operation
+{
+    kOperationRead,
+    kOperationWrite
+};
+
+static bool Invoke_REAPI_Cache_Client(const HashDigest& digest, StatCache *stat_cache, const FrozenArray<FrozenFileAndHash>& outputFiles, ThreadState* thread_state, Operation operation, const Frozen::DagNode* dagNode, Mutex* queue_lock)
+{
+    ProfilerScope profiler_scope("InvokeCacheMe", thread_state->m_ProfilerThreadId, outputFiles[0].m_Filename);
+
+    const char* reapi_raw = getenv(kENV_REAPI_CACHE_CLIENT);
+    if (reapi_raw == nullptr)
+        return false;
+
+    PathBuffer pathbuf;
+    PathInit(&pathbuf, reapi_raw);
+    char reapi[kMaxPathLength];
+    PathFormat(reapi, &pathbuf);
+
+    char buffer[5000];
```

Comment:
Are we ok with keeping this fixed-size string here? It does seem very unlikely that we would overrun it (would take a very loong `kENV_REAPI_CACHE_CLIENT` string), but it still seems like bad practice. Let's either make this a variable length string, or at least check and error out if `totalWritten` > string size.

Reply:
fixed